posts: announce rustup 1.28.2 (#1593)
rust-lang/blog.rust-lang.org: content/Rustup-1.28.2.md

Reviewer: I am aware that we have a call-for-testing post in <https://internals.rust-lang.org/t/seeking-beta-testers-for-rustup-v1-28-2>. Do we want to call for testing also in the "Inside Rust Blog"? I could see this is a patch version release so we skip that part :)

Reviewer 2: Oh it might also be worth mentioning this on zulip. One of the things brought up before is that many people involved with rust don't read internals whereas zulip is more active.

Author: @weihanglo That might indeed be a good idea in general, but the scope of this patch release is relatively small IMHO.

Reviewer 2: @rami3l We should post in general for visibility I think because it should have a wide audience.

Reviewer 2: Perhaps it should link to a t-rustup thread (or vice versa) so there's a cross-reference of where rustup discussions have happened on zulip.

Author: @ChrisDenton Already posted to [#general > Seeking beta testers for rustup v1.28.2](https://rust-lang.zulipchat.com/#narrow/channel/122651-general/topic/Seeking.20beta.20testers.20for.20rustup.20v1.2E28.2E2).

+++
path = "2025/04/25/Rustup-1.28.2"
title = "Announcing rustup 1.28.2"
authors = ["The Rustup Team"]
aliases = ["2025/04/25/Rustup-1.28.2.html"]
+++

The rustup team is happy to announce the release of rustup version 1.28.2.
[Rustup][install] is the recommended tool to install [Rust][rust], a programming language that
empowers everyone to build reliable and efficient software.

## What's new in rustup 1.28.2

The headlines of this release are:

- The cURL download backend and the native-tls TLS backend are now officially deprecated and
  a warning will start to show up when they are used. [pr#4277]

  - While rustup predates reqwest and rustls, the rustup team has long wanted to standardize on
    an HTTP + TLS stack in Rust, which should increase security, potentially improve performance, and
    simplify maintenance of the project.
    With the default download backend already switched to reqwest since [2019][pr#1660], the team
    thinks it is time to focus maintenance on the Rust-based stack.

>   - The rustup team encourages everyone to switch to the reqwest backend with rustls, and would love

Reviewer 3: Presumably there's no actual work involved here in this encouragement, right? This is just the default with the new release? Can we rephrase this to make that clear?

    reqwest and rustls, and would love to hear from you about your use case via

>     [GitHub Issues][issue tracker] if it does not work well with your particular setup.

Reviewer 4: I think this could be clearer still, maybe:

"For people who have RUSTUP_USE_CURL=1 or RUSTUP_USE_RUSTLS=0 set in their environment to work around issues with rustup, please try to unset these after upgrading to 1.28.2 and file an issue [..] if you still encounter problems."

Author: Fixed!

+++

> path = "2025/04/25/Rustup-1.28.2"

Reviewer 3: ```suggestion
path = "2025/05/05/Rustup-1.28.2"
```

title = "Announcing rustup 1.28.2"
authors = ["The Rustup Team"]

> aliases = ["2025/04/25/Rustup-1.28.2.html"]

Reviewer 3: ```suggestion
aliases = ["2025/05/05/Rustup-1.28.2.html"]
```

+++

The rustup team is happy to announce the release of rustup version 1.28.2.

> [Rustup][install] is the recommended tool to install [Rust][rust], a programming language that is
> empowering everyone to build reliable and efficient software.

Reviewer 5: ```suggestion
[Rustup][install] is the recommended tool to install [Rust][rust], a programming language that 
empowers everyone to build reliable and efficient software.
```

## What's new in rustup 1.28.2

The headlines of this release are:

- The cURL download backend is now officially deprecated and a warning will start to show up
  when it is used. [pr#4277]

>   - While rustup predates reqwest and rustls, the rustup team has long wanted to standardize on

Reviewer 5: ```suggestion
  - The rustup team has long wanted to standardize on
```

Reviewer 4: This seems like useful context to explain why we didn't start out with reqwest/rustls?

Reviewer 5: Maybe, but for a notification post like this, I'm not so worried about why the *previous* decision was made but more interested in what's changing, what I need to do about it, and then maybe why the change is happening. If historical context is needed, I'd hope it could come after the other points have been addressed.

Author: I agree with @djc with this, so I guess it'd still be better to keep it here 🙏

>     an HTTP + TLS stack in Rust, which should increase security, potentially improve performance, and

Reviewer 5: Is the actual stack now all in Rust? I see `native-tls` in the Cargo.lock file, which presumably is non-Rust code.

Reviewer 4: It is not -- reqwest + rustls is now the default but in addition to the cURL fallback there is still a native-tls fallback option.

Author: @shepmaster Thanks! I think I can change the ending to "... maintaining the reqwest-based stack". A Rust-based stack is our long-term goal but probably it's too early to say that right now.

Reviewer 5: > there is still a native-tls fallback

Is that intended to be deprecated eventually as well? If not, then it seems like the overall goal isn't "there is only a pure-Rust stack". Since the benefit of "a pure-Rust stack by default" has already occurred, it seems that this change is purely about reducing maintenance burden (still a good thing!).

Author: @shepmaster IIRC the deprecation of `native-tls` is indeed on our current project roadmap.

Reviewer 5: Depending on how far out you think that may be, it could be worth a tiny mention here and now. That way, people don't jump from cURL to native-tls only to need to switch again, at least without trying the intended stack first (and filing bugs if needed 🤞)

Reviewer 4: Fair point; https://github.com/rust-lang/rustup/pull/4317.

Author: @djc I have mirrored your change in this PR; I'll soon do the same in https://github.com/rust-lang/rustup/pull/4313.

Author: @Mark-Simulacrum Good point! I've adjusted the phrasing.

    simplify maintenance of the project.
    With the default download backend already switched to reqwest since [2019][pr#1660], the team
    thinks it is time to start removing the cURL backend and focus on maintaining the Rust-based
    stack.

  - The rustup team encourages everyone to switch to the reqwest backend, and would love to hear from

>     you about your use case via [GitHub Issues][issue tracker] if it does work against your particular

Reviewer 5: Does *not* work, perhaps?

Author: @shepmaster The negation is in "against", but I guess it's a bit too easy to misread? Fixed anyway :)

Reviewer 5: Yes, I would definitely have never read "against" as a negation (and I still don't, even with your response). For example, "I tested my changes against the main branch" has no such negative meaning to me.

    particular setup.

> - Version pinning of `rustup` is now supported when installing it via `rustup-init.sh` with the `RUSTUP_VERSION`
>   environment variable set to e.g. `1.28.2`; the same configuration will be respected when performing
>   `rustup self update` from this version on, enabling arbitrary downgrades in the meantime. [pr#4259]

Reviewer 5: ```suggestion
- The version of `rustup` can be pinned when installing via `rustup-init.sh`. To do so, set the `RUSTUP_VERSION` environment variable to the desired version (for example `1.28.2`). 
```

Are there other ways of pinning the version, such as if I've already installed, or *only* via this installation method?

> the same configuration will be respected when performing `rustup self update` from this version on

I don't know what this means. Guessing, it means that `rustup self update` will no longer actually perform an update. I hope that rustup prints a warning to the user that the version is pinned to X at that point and describes how to unpin/change the pin. Guessing further, I assume that you can change the pin by setting the environment variable again when running `rustup self update`.

> enabling arbitrary downgrades in the meantime. 

I don't know what this means. My first reading of this is "the version of rustup with decrease at some point, whenever rustup decides it". I cannot believe that is the actual intention, however. "Arbitrary" is not a feature I look for in my development tools, especially paired with a nebulous timeframe like "meantime".
